pkg/orm/client.go

```diff
@@ -0,0 +1,610 @@
+		return nil, cerrors.ErrMetaOpFail.Wrap(err)
+	}
+
+	db.SetConnMaxIdleTime(conf.ConnMaxIdleTime)
+	db.SetConnMaxLifetime(conf.ConnMaxLifeTime)
+	db.SetMaxIdleConns(conf.MaxIdleConns)
+	db.SetMaxOpenConns(conf.MaxOpenConns)
+	return db, nil
+}
+
+func newClient(sqlDB *sql.DB) (*metaOpsClient, error) {
+	db, err := gorm.Open(mysql.New(mysql.Config{
+		Conn:                      sqlDB,
+		SkipInitializeWithVersion: false,
+	}), &gorm.Config{
+		SkipDefaultTransaction: true,
+		// TODO: logger
+	})
+	if err != nil {
+		log.L().Error("create gorm client fail", zap.Error(err))
+		return nil, err
+	}
+
+	return &metaOpsClient{
+		db:   db,
+		impl: sqlDB,
+	}, nil
+}
+
+// metaOpsClient is the meta operations client for framework metastore
+type metaOpsClient struct {
+	// gorm claim to be thread safe
+	db   *gorm.DB
+	impl *sql.DB
+}
+
+func (c *metaOpsClient) Close() error {
+	if c.impl != nil {
+		return c.impl.Close()
+	}
+
+	return nil
+}
+
+////////////////////////// Initialize
+// Initialize will create all related tables in SQL backend
+// TODO: What if we change the definition of orm??
+func (c *metaOpsClient) Initialize(ctx context.Context) error {
+	if err := c.db.AutoMigrate(&model.ProjectInfo{}, &model.ProjectOperation{}, &libModel.MasterMetaKVData{},
+		&libModel.WorkerStatus{}, &resourcemeta.ResourceMeta{}, &model.LogicEpoch{}); err != nil {
+		return cerrors.ErrMetaOpFail.Wrap(err)
+	}
+
+	// check first record in logic_epochs
+	return c.InitializeEpoch(ctx)
+}
+
+/////////////////////////////// Logic Epoch
+// TODO: what if the record is deleted manually??
+func (c *metaOpsClient) InitializeEpoch(ctx context.Context) error {
+	var logicEp model.LogicEpoch
+	// first check and add first record if not exists
+	if result := c.db.First(&logicEp, defaultEpochPK); result.Error != nil {
+		if result.Error == gorm.ErrRecordNotFound {
+			if res := c.db.Create(&model.LogicEpoch{
+				Model: model.Model{
+					SeqID: defaultEpochPK,
+				},
+				Epoch: defaultMinEpoch,
+			}); res.Error != nil {
+				return cerrors.ErrMetaOpFail.Wrap(res.Error)
+			}
+
+			return nil
+		}
+
+		return cerrors.ErrMetaOpFail.Wrap(result.Error)
+	}
+
+	// already exists, do nothing
+	return nil
+}
+
+func (c *metaOpsClient) GenEpoch(ctx context.Context) (libModel.Epoch, error) {
+	var epoch libModel.Epoch
+	err := c.db.Transaction(func(tx *gorm.DB) error {
+		//(1)update epoch = epoch + 1
+		if err := tx.Model(&model.LogicEpoch{
+			Model: model.Model{
+				SeqID: defaultEpochPK,
+			},
+		}).Update("epoch", gorm.Expr("epoch + ?", 1)).Error; err != nil {
+			// return any error will rollback
+			return err
+		}
+
+		//(2)select epoch
+		var logicEp model.LogicEpoch
+		if err := tx.First(&logicEp, defaultEpochPK).Error; err != nil {
+			return err
+		}
+		epoch = libModel.Epoch(logicEp.Epoch)
+
+		// return nil will commit the whole transaction
+		return nil
+	})
+	if err != nil {
+		return libModel.Epoch(0), cerrors.ErrMetaOpFail.Wrap(err)
```

Comment:
Retry here?

Reply:
I want the retry logic be a upper layer.
It will do it later.